[WIP][Do Not Merge] Add initial dev guide (#47)
urbit/tlon.io: content/dev/dev-guide.md

# Tlon Dev Guide
## What is this guide?
This is Tlon's guide for new developers that want to build applications on Urbit. The goal of this guide is to take you from a fresh machine to a working development environment for Urbit applications. This guide assumes macOS, but Linux is also supported (see the Urbit foundation docs for Linux support) It's important this guide remains clear and correct, if you find something broken please let us know at support@tlon.io

## Pre-requisites

### Homebrew
- Homebrew is a package manager for macOS, we'll need it installed in order to install `git-lfs`.
- ```/bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)"```

### Git LFS
- A package that enables git to work with large files, we'll install this via brew.
- ```brew install git-lfs```

### Wget
 - A package used to download artifacts from the web

>  - `brew install wget`

Reviewer: Doesn't `curl` come preinstalled on most systems? Shouldn't we use that instead? Also the instructions below actually use `curl`, so...

Author: Yeah, the first part used curl and a later part in the docs used wget which is why I added this. To your point I think just standardizing on curl makes sense and reduces steps.

### Visual Studio Code
 - While not technically a prerequisite, there's a hoon syntax highlighting plugin available and this is the editor we recommend: https://code.visualstudio.com/
 - For the small text file edits in this guide, we'll simply use `nano`.

### Urbit Install
 - First we need to create a directory for the urbit binary, we'll create this in your home directory.
   - ```bash
     mkdir ~/urbit-dev
     ```
 - Next we're going to pull down the latest urbit binary from urbit.org from the newly created directory.
    - ```bash
      cd ~/urbit-dev
      curl -JLO https://urbit.org/install/mac/latest
      ```
  - Now we'll untar the contents of the download into our newly created directory.
    - ```bash 

>       tar zxvf ./darwin.tgz --strip=1

Reviewer 2: ```suggestion
      tar -zxvf ./darwin.tgz --strip=1
```

Author: I think you don't need the `-` for tar, but I've added it anyway since it works either way.

Reviewer 2: TIL, cool

      ```
  - With the urbit binary installed we're ready for the basic dev workflow.

## Development Basics

### Fake Ships
#### What is a "fake" ship?

>  - For local development we'll create a 'fake' ship, this is a local ship that is not connected to the urbit network. This means we don't need a real ID (eg comet, planet) and can delete and recreate the ship whenever we want to reset things.

Reviewer: Pretty sure house style is to always put whitelines around headers.

```suggestion
## Development Basics

### Fake Ships

#### What is a "fake" ship?

 - For local development we'll create a 'fake' ship, this is a local ship that is not connected to the urbit network. This means we don't need a real ID (eg comet, planet) and can delete and recreate the ship whenever we want to reset things.
```

Probably forgivable for the `####` ones because they get so small/granular, but check for consistency with all the other docs.

#### Creating a fake ship
 - We can create a local fake ship with the below command. This will create a fake ship named zod in our `urbit-dev` directory. This will take a couple of minutes and generate some output in the terminal.
   - ```bash

>      ./urbit -F zod

Reviewer: Note that this just uses the livenet pill and the files that come with it. That might not include the latest updates.

Author: Yeah - I may want to restructure this a bit so it's not repetitive, but it's also cleaner to explain it this way. I think it'd be helpful for me to understand more about pills.

     ```
 - When it's finished you'll see a command prompt like `~zod:dojo>` - this is the urbit dojo where you can enter and run urbit commands.
 - To exit the dojo type `|exit`
#### Fake ship networking
 - Fake ships on the same machine can automatically talk to each other. Let's create a second fake ship named `~bus` and confirm this.
   - ```bash
     ./urbit -F bus
     ```
 - Now you'll have two local ships, `~zod` and `~bus`.
 - From the newly created `~bus` dojo enter `|hi ~zod` and you'll see that urbit's networking finds the first fake ship we created.
 - Exit the `~bus` dojo with `|exit`

#### Faster fake ship booting

> A common part of the development workflow is deleting and creating fake ships. The couple of minutes this takes slows down development. Instead we can create a backup copy of a freshly created fake ship. This way when you want to delete some fake ship you've been working on and reset it, it can be done instantly rather than having to wait for a new one to be created.

Reviewer: > common

This really depends on what you're working on. I've ran fake ships into the upper hundreds of clay revisions while just doing userspace development. If you only need to do it once a day or less, booting a fresh one really isn't all that time-consuming.

Using the same backup for a long time might also mean working with files that are way behind latest. Just things to be aware of.

Author: This is good to know and I think changes my mind about this?

If the in the common case you don't have to boot a new fake ship that often, then we could reduce a lot of the guide (move this into an advanced tips and tricks section rather than the default).

 - We'll do this now by backing up the fake `~zod` we just created.
 - First start up `~zod` from our `~/urbit-dev` directory. Notice that we don't need the `-F` argument this time because `~zod` is already created.
   - ```bash
     ./urbit zod
     ```
 - Next mount every desk (this is so we don't have to do this every time after resetting our fake ship). Each of these is one dojo command.
   - ```bash
     |mount %base
     |mount %garden
     |mount %landscape
     |mount %bitcoin
     |mount %webterm
     ```

>  - Next shutdown `~zod` with `|exit`

Reviewer: ```suggestion
 - Next shutdown `~zod` with `|exit` or ctrl-D
```

     ```
 - Then you'll dev with `~zod` like normal, but when you're ready to reset it to a fresh state, instead of creating a new one we'll just delete `~zod` and copy the backup over.
   - ```bash
     rm -r zod
     cp -r zod.bk zod
     ```
 - This will create a new `zod` directory and you can run it as usual with `./urbit zod`.

### Desks
#### What is a desk?
A desk is an independently revision-controlled branch of a ship that uses the Clay filesystem. Each desk contains its own apps, mark definitions, files, etc. You can think of a desk like a git branch you're working in.

A basic ship has a `%base`, `%garden`, `%landscape`, `%bitcoin`, and `webterm` desk (we mounted these earlier). The `%base` desk has the kernel and base system software, while other desks are usually each for different apps. 

A desk is a series of numbered commits, the most recent of which represents the current state of the desk. A commit is composed of an absolute time when it was created, a list of zero or more parents, and a map from paths to data.

For building applications to distribute on urbit, we'll want to work on a new desk.

#### Desk Basics

##### Mounting and Unmounting Desks

>  - Mounting a desk allows you to add files to it (it creates a directory you can see on your mac). Run these commands in the dojo.

Reviewer: ```suggestion
 - Mounting a desk allows you to add files to it (it creates a directory you can see on the host filesystem). Run these commands in the dojo.
```

   - ```bash
     |mount %base
     ```
 - You can similarly unmount a desk
   - ```bash
     |unmount %base
     ```
##### Setting Up a New Desk

>  - To create a new desk we'll need to merge from an existing one. Typically this will be `%base`, think of this like creating a new branch. A simple example of this command is below, but don't enter it yet - we'll need to do something else first.

Reviewer: > but don't enter it yet - we'll need to do something else first.

Come on, don't give me (the theoretical reader) this command then. My reading comprehension is like 50% so I might miss this!

But also it doesn't really matter, right?

Author: Yeah this is a good flag, it was an artifact of me working through the guides in order

   - ```bash
     |merge %mydesk our %base
     ```
 - Every new desk needs a number of base files, in order to simplify this a bit we've created a script that symlinks all the files necessary for interacting with the `%base` and `%garden` desks.
 - In order to use this we'll first want to clone down a local copy of the `urbit` code.
   - ```bash
     git clone https://github.com/urbit/urbit ~/urbit-git
     ```
 - In `urbit-git/pkg` we can see `base-dev` and `garden-dev` desks. These desks contain symlinks to all of the `/sur`, `/lib` and `/mar` files necessary for interacting with the `base` and `garden` desks. Don't worry about these details for now, but we'll be using these to set up our initial environment.
 - Let's create a new `hello` desk.
 - First navigate to the `pkg` directory in `urbit-git`
   - ```bash
     cd ~/urbit-git/pkg
     ```
 - Next create a `hello` directory
   - ```bash
     mkdir hello

>      ```

Reviewer: > creating your desk within the official repo

This is bad and it will make devs feel bad. Does the symbolic-merge script work for targets outside the repo? Probably not. Should it? Maybe. Can we do anything else in the mean time? Surely yes.

My personal flow is to just copy `base-dev` and/or `garden-dev` contents into the target desk on-demand, which is usually once during development (if even that, you can just start with `|merge %mydesk our %garden` and then you have it all), and once during initial deploy. The advantage here is that files you didn't touch don't magically change the next time you do a copy. The disadvantage, too, is that that doesn't happen: you might miss an API change, only find out at run-time.

Author: I need to ask more about this - I wrote up the existing instructions but don't know enough about what's going on here to reason about them. I suspect your method is better and this bit I think is the part people currently complain the most about?

Author: Ok, I've reworked this based on our discussion and think the new approach (primarily from your comment here) is much less complicated and easier to understand.

 - Next run the included `symbolic-merge.sh` script from the `pkg` directory.
   - ```bash
     ./symbolic-merge.sh base-dev hello
     ./symbolic-merge.sh garden-dev hello
     ```
 - Now we should see `lib`, `mar`, and `sur` inside `hello`.
 - Our desk must include a `sys.kelvin` file which specifies the kernel version it's compatible with. Let's create that.
   - ```bash

>      echo "[%zuse 420]" > sys.kelvin

Reviewer: ```suggestion
     echo "[%zuse 419]" > sys.kelvin
```

Latest!

     ```
 - We can also add a `desk.ship` file to specify the original publisher of this desk. Since we're going to use our fake ship `~zod` let's add `~zod` as the publisher.
   - ```bash
     echo "~zod" > desk.ship

>      ```

Reviewer: Notably optional. Also we don't really sign releases or anything rn, so no real way to verify.

Author: I can add a note about both of these things for this.

 - If we had Gall agents in this desk which should be automatically started when the desk is installed, we'd add them to a hoon list in the `desk.bill` file. It would look something like this:

>    - ```bash

Reviewer 2: We should actually get the Hoon highlighter in here. I can help with that.

   - ```bash
     nano desk.docket-0
     ```
   - In the text editor paste the following and save:
   - ```bash
     :~  title+'Hello'
         info+'A simple hello world app.'
         color+0x81.88c9
         image+'https://media.urbit.org/docs/userspace/dist/wut.svg'
         base+'hello'
         glob-ames+[~zod 0v0]
         version+[0 0 1]
         website+'https://urbit.org/docs/userspace/dist/guide'
         license+'MIT'
     ==
     ```
  - This is a `docket` file. This file sets various options for the desk with a tile and usually a brower-based front-end of some kind. 
  - It mainly configures the appearance of an app's tile, the source of its glob (bundle of client resources like HTML, JS, and CSS files), and some additional metadata.
    - We've given the app a `%title` of "Hello", which will be displayed on the app tile and will be the name of the app when others browse to install it. We've given the app tile a `%color` of #8188C9, and also specified the URL of an `%image` to display on the tile.
    - The `%base` clause specifies the base URL path for the app. We've specified "hello" so it'll be `http://localhost:80/apps/hello/...` in the browser. 

>     - For the glob, we've used a clause of `%glob-ames`, which means the glob will be served from a ship over Ames, as opposed to being served over HTTP with a `%glob-http` clause or having an Eyre binding with a `%site clause`. 

Reviewer: Note that `%base` only gets put into `/apps/[base]` _because_ the app is served as a glob through docket. A `%site` would be free to bind wherever. Not sure if the `%base` value gets used anywhere else, but don't think so.

Reviewer: Also the `:80` is dependent on urbit being able to bind to that. It might already be in use by something else, for example. Probably worth pointing out the in-use port can be found among the startup messages.

>     - For the glob, we've used a clause of `%glob-ames`, which means the glob will be served from a ship over Ames, as opposed to being served over HTTP with a `%glob-http` clause or having an Eyre binding with a `%site clause`. 

Reviewer 2: The sooner we can kill globs the better. So much implied information in this one sentence. Perhaps we can link to some of the urbit.org explainers for these?

    - Next we've specified `[~zod 0v0]`. Since `~zod` is the fakeship we'll install it on, the `%docket` agent will wait for a separate upload of the glob, so we can just specify `0v0` here as it'll get overwritten later.
    - The `%version` clause specifies the version as a triple of major version, minor version and patch version. The rest is just some additional informative metadata which will be displayed in App Info.

> - With that saved we now have everythign we need in the `hello` directory.

Reviewer 2: ```suggestion
- With that saved we now have everything we need in the `hello` directory.
```

##### Installing The New Desk Into Fake Zod
- First we're going to create a new fake zod that's bootstrapped with a pill from our urbit-git repo. The pills are stored in git lfs and need to be pulled into our repo first.
  - ```bash
    cd ~/urbit-git
    git lfs install
    git lfs pull
    ```
- Next we're going to create the fake zod using the `multi-brass.pill` we just pulled down.
  - ```bash
    cd ~/urbit-dev
    ./urbit -F zod -B ~/urbit-git/bin/multi-brass.pill

>     ```

Reviewer: Consider using the solid `multi.pill` for faster boot times.

Author: What's the difference?

Reviewer: Solid pills are precompiled. Brass pills recompile everything during boot.

Author: I added a section explaining pills.

- Once our new fake zod is booted we'll create and mount a new `%hello` desk in dojo.
  - ```bash
    |merge %hello our %base
    |mount %hello
    ```
- We should now see `hello` inside our `zod` directory.
- Currently it's just a clone of `%base` so let's delete its contents.
  - ```bash
    cd ~/urbit-dev/zod
    rm -r hello/*
    ```
- Next we'll copy the contents of the `hello` desk we created earlier. We'll need to use `cp -LR` to resolve all the symlinks.
  - ```bash
    cp -LR ~/urbit-git/pkg/hello/* hello/
    ```
- Back in the `zod` dojo we can commit the changes and install the desk.
  - ```bash
    |commit %hello
    |install our %hello
    ```
  - You'll see `docket: awaiting manual glob for %hello desk` this is because our `desk.docket-0` files includes a `%glob-ames` clause which specifies our ship as the source, so it's waiting for us to upload the glob.

> - Open a browser and navigate to `http://localhost:80` and login with the default fake zod code `lidlut-tabwed-pillex-ridrup`. You'll see our tile, but it says installing with a spinner due to the missing glob.

Reviewer: Might be better to teach them to run `+code` to retrieve the code instead?

##### Create Files for Glob
We'll now create the files for the glob. We'll use a very simple static HTML page that just displayes "Hello World!" and an image. Typically we'd have a more complex JS web app that talked to apps on our ship through Eyre's channel system, but for the sake of simplicity we'll forgo that. Let's hop back in the Unix terminal.
 - ```bash
   cd ~
   mkdir hello-glob
   cd hello-glob
   mkdir img
   wget -P img https://media.urbit.org/docs/userspace/dist/pot.svg
   ```
- We've grabbed an image to use in our "Hello world!" page. The next thing we need to add is an `index.html` file in the root of the folder. The `index.html` file is mandatory; it's what will be loaded when the app's tile is clicked. Let's open our preferred editor and create it.
  - ```bash
    nano index.html
    ```
- Paste the following into the editor and save it.

>   - ```html
>     <!DOCTYPE html> <html> <head> <style> div { text-align: center; }
>     </style> </head> <title>Hello World</title> <body> <div>
>     <h1>Hello World!</h1> <img src="img/pot.svg" alt="pot" width="219"
>     height="196" /> </div> </body> </html>
>     ```

Reviewer: The `<title>` probably shouldn't live outside the `<head>`. Everything else here is just a stylistic change, but should help make this slightly less opaque.

```suggestion
  - ```html
    <!DOCTYPE html>
    <html>
      <head>
        <title>Hello World</title>
        <style>body { text-align: center; }</style>
      </head>
      <body>
        <h1>Hello World!</h1>
        <img src="img/pot.svg" alt="pot" width="219" height="196" />
      </body>
    </html>
    ```
```

Reviewer: Also I feel like we shouldn't lose the nuance of docket's glob capability mostly being a loom limits workaround. Questions like "why not copy this in through clay" will inevitably arise, and for this specific example that's totally reasonable (no big js blobs, after all).

Author: This nuance was not clear to me from the existing guide - it sounds like a good note to add, but I need to ask you more about it in order to understand it well enough to write it up clearly.

Author: Added context based on 1:1 with Mark

##### Upload to Glob
We can now create a glob from the directory. To do so, navigate to http://localhost:80/docket/upload in the browser. This will bring up the %docket app's Globulator tool.

- Simply select the `hello` desk from the drop-down, click `Choose file` and select the `hello-glob` folder in the the file browser, then hit `glob!`.
- Now if we return to our ship's homescreen, we should see the tile looks as we specified in the docket file: http://localhost:80
- And if we click on the tile, it'll load the index.html in our glob.
- Our app is working!
##### Publish
The final step is publishing our desk with the `%treaty` agent so others can install it. To do this, there's a simple command in the dojo.
 - ```bash
   :treaty|publish %hello
   ```

>    - Note: For desks without a docket file (and therefore without a tile and glob), treaty can't be used. Instead you can make the desk public with `|public %desk-name`.

Reviewer: Should clarify further and explain that desks aren't publicly readable by default, and that `:treaty|publish` not only publishes the metadata, but also makes the desk publicly readable, like `|public`.

##### Remote Install
Let's spin up another fake ship so we can try to install our newly published app.

 - ```bash
   cd ~/urbit-dev
   ./urbit -F bus
   ```
   - Note: For desks without a docket file (and therefore without a tile and glob), users cannot install them through the web interface. Instead remote users can install it from the dojo with `|install ~our-ship %desk-name`.

> - In the browser, navigate to http://localhost:8081 and login with `~bus`'s code `riddec-bicrym-ridlev-pocsef`. Next, type `~zod/` in the search bar, and it should pop up a list of `~zod`'s published apps, which in this case is our `Hello` app.

Reviewer: This is likely going to be on port `8080`, if ~zod bound successfully to `80`. Again, ymmv.

Author: It was on `80` when I ran through the guide myself (hence the change), but both these notes are good. I can explicitly state this and note that the bound port is present in the startup output.

## Development Basics

### Fake Ships
#### What is a "fake" ship?
 - For local development we'll create a 'fake' ship, this is a local ship that is not created to the urbit network. This means we don't need a real ID (eg comet, planet) and can delete and recreate the ship whenever we want to reset things.
#### Creating a fake ship
 - We can create a local fake ship with the below command. This will create a fake ship named zod in our `urbit-dev` directory. This will take a couple of minutes and generate some output in the terminal.
   - ```bash
     ./urbit -F zod
     ```
 - When it's finished you'll see a command prompt like `~zod:dojo>` - this is the urbit dojo where you can enter and run urbit commands.
 - To exit the dojo type `|exit`
#### Fake ship networking
 - Fake ships on the same machine can automatically talk to each other. Let's create a second fake ship named `~bus` and confirm this.
   - ```bash
     ./urbit -F bus
     ```
 - Now you'll have two local ships, `~zod` and `~bus`.
 - From the newly created `~bus` dojo enter `|hi ~zod` and you'll see that urbit's networking finds the first fake ship we created.

>  - Exit the `~bus` dojo with `|exit`

Reviewer 3: ```suggestion
 - Exit the `~bus` dojo with `|exit` or ctrl-D.
 ```

##### Create Files for Glob
We'll now create the files for the glob. We'll use a very simple static HTML page that just displayes "Hello World!" and an image. Typically we'd have a more complex JS web app that talked to apps on our ship through Eyre's channel system, but for the sake of simplicity we'll forgo that. Let's hop back in the Unix terminal.
 - ```bash
   cd ~
   mkdir hello-glob
   cd hello-glob
   mkdir img
   wget -P img https://media.urbit.org/docs/userspace/dist/pot.svg
   ```
- We've grabbed an image to use in our "Hello world!" page. The next thing we need to add is an `index.html` file in the root of the folder. The `index.html` file is mandatory; it's what will be loaded when the app's tile is clicked. Let's open our preferred editor and create it.
  - ```bash
    nano index.html
    ```
- Paste the following into the editor and save it.
  - ```html
    <!DOCTYPE html> <html> <head> <style> div { text-align: center; }
    </style> </head> <title>Hello World</title> <body> <div>
    <h1>Hello World!</h1> <img src="img/pot.svg" alt="pot" width="219"
    height="196" /> </div> </body> </html>
    ```
##### Upload to Glob

> We can now create a glob from the directory. To do so, navigate to http://localhost:80/docket/upload in the browser. This will bring up the %docket app's Globulator tool.

Reviewer 3: ```suggestion
We can now create a glob from the directory. To do so, navigate to http://localhost:80/docket/upload in the browser. This will bring up the `%docket` app's Globulator tool.
```

# Tlon Dev Guide

## What is this guide?

> This is Tlon's guide for new developers that want to build applications on Urbit. The goal of this guide is to take you from a fresh machine to a working development environment for Urbit applications. This guide assumes macOS, but Linux is also supported (see the Urbit foundation docs for Linux support) It's important this guide remains clear and correct, if you find something broken please let us know at support@tlon.io

Reviewer 4: "This is a guide for new developers _who_ want to build applications on Urbit."

Period needed after parenthetical "...for Linux support)."

Recommend direct link to email because last sentence needs a period at the end (but it might be confusing with the address).

## Pre-requisites

### Homebrew

> - Homebrew is a package manager for macOS, we'll need it installed in order to install `git-lfs`.

Reviewer 4: Replace comma with semicolon (or period and split into two sentences).

- ```/bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)"```

### Git LFS

> - A package that enables git to work with large files, we'll install this via brew.

Reviewer 4: Start bullet with "Git LFS is..." to avoid awkward syntax.

Comma should be a semicolon (or another sentence completely).

- ```brew install git-lfs```

### Visual Studio Code

>  - While not technically a prerequisite, there's a hoon syntax highlighting plugin available and this is the editor we recommend [Visual Studio Code](https://code.visualstudio.com/).

Reviewer 4: "While it's not technically required, you can install a hoon syntax highlighting plugin. This is the editor we recommend."

 - For the small text file edits in this guide, we'll simply use `nano`.

### The Urbit Git repository

  - We're going to need a local copy of the Urbit source code.
     - ```bash
        git clone https://github.com/urbit/urbit ~/urbit-git
       ```
  - Pull down the files stored using git lfs. 
    - ```bash
      cd ~/urbit-git
      git lfs install
      git lfs pull
      ```

### The Urbit binary

>  - First we need to create a directory for the urbit binary, we'll create this in your home directory.

Reviewer 4: Replace comma with semicolon.

   - ```bash
     mkdir ~/urbit-dev
     ```

>  - Next we're going to pull down the latest urbit binary from urbit.org from the newly created directory.

Reviewer 4: This sentence could be clarified a bit.

    - ```bash
      cd ~/urbit-dev
      curl -JLO https://urbit.org/install/mac/latest
      ```
  - Now we'll untar the contents of the download into our newly created directory.
    - ```bash 
      tar -zxvf ./darwin.tgz --strip=1
      ```

>   - With the urbit binary installed we're ready for the basic dev workflow.

Reviewer 4: Comma after installed

## Development basics

### Fake ships

#### What is a "fake" ship?

>  - For local development we'll create a 'fake' ship, this is a local ship that is not connected to the Urbit network. This means we don't need a real ID (eg comet, planet) and can delete and recreate the ship whenever we want to reset things.

Reviewer 4: Would specify real ID as "real Urbit ID"

eg should be e.g.

"...and can delete _or_ recreate the ship..."

#### What is a "pill"?

 - A pill file contains a sequence of events that initializes an urbit ship (its boot sequence). If no pill is specified, the livenet pill is downloaded instead. 
 - Since OS updates propagate through the network, it's possible the default pill file could be slightly out of date.

>  - We can manually get the most up to date pill file for local development by cloning down and using the up to date version in the urbit source code repo when starting our fake ship.

Reviewer 4: up-to-date is hyphenated

>    - For fake development ships not talking to the regular network this works well.

Reviewer 4: Comma after network.

   - We don't do this for 'real' ships on the actual network because if you're using a pill that's newer than what your sponsor has (the ship that you get updates from) your sponsor will think you're on an older version.
   - When this happens your sponsor will attempt to upgrade you, but will actually be downgrading you to their verison.
   - There are compiled and uncompiled pills. They are differentiated by a naming convention that adds `-brass` to the compiled pill name. For example, `multi.pill` is the name of a compiled pill and the name of its uncompiled version is `multi-brass.pill`. For the purposes of this guide we'll use compiled pills because they boot faster. An uncompiled pill lets you audit the source code if you need to.

>  - In summary, if you're using a fake ship -> get the latest pill, if you're starting a real ship don't worry about pills.

Reviewer 4: Comma after pill should be a period or a semicolon.

   - ```bash
     ./urbit -F bus -B ~/urbit-git/bin/multi.pill
     ```
 - Now you'll have two local ships, `~zod` and `~bus`.
 - From the newly created `~bus` dojo enter `|hi ~zod` and you'll see that urbit's networking finds the first fake ship we created.
 - Exit the `~bus` dojo with `|exit` or `ctrl-D`.

### Desks

#### What is a desk?

A desk is an independently revision-controlled branch of a ship that uses the Clay filesystem. Each desk contains its own apps, type definitions, files, etc. You can think of a desk like a git branch you're working in.

A basic ship has a `%base`, `%garden`, `%landscape`, `%bitcoin`, and `webterm` desk. The `%base` desk has the kernel and base system software, while other desks are each for different apps.

A desk is a series of numbered commits, the most recent of which represents the current state of the desk. A commit is composed of an absolute time when it was created, a list of zero or more parents, and a map from paths to data.

For building applications to distribute on Urbit, we'll want to work on a new desk.

#### Desk basics

##### Mounting and unmounting Desks

>  - Mounting a desk allows you to add files to it (it creates a directory you can see on the host filesystem). Run these commands in the dojo.

Reviewer 4: Suggest eliminating parenthetical and instead writing "Mounting a desk allows you to add files by creating a directory that you can see on the host filesystem."

   - ```bash
     |mount %base
     ```
 - You can similarly unmount a desk
   - ```bash
     |unmount %base
     ```

##### Setting up a new desk

>  - To create a new desk we'll need to merge from an existing one, think of this like creating a new branch.

Reviewer 4: Comma after one should be a period / new sentence.

 - Let's create a new `hello` desk by merging from `%garden` (`garden` is just an old name for the application now referred to as `landscape`).
  - Let's start up our fake `zod`.
    - ```bash
      cd ~/urbit-dev
      ./urbit zod
      ```
  - Create our desk in the dojo by merging from `%garden`
    - ```bash
      |merge %hello our %garden
      ```
 - **Note**: Thinking about dependencies:
   - When we merge from `%garden` we start with its base files in our application.

>    - Since we're copying these files over, its possible an API change could happen in `urbit-git` while we're building our app.

Reviewer 4: Change its to it's

>    - It's good practice to run the merge command at the start of development and prior to deployment (after a fresh pull of `urbit-git`) to make sure it works with the up to date system changes.

Reviewer 4: Up-to-date is hyphenated.

  - Mount `%hello` so we can access it from the host filesystem.
    - ```bash
      |mount %hello
      ```
  - Exit the Dojo with `|exit` or `ctrl-D`.
 - Now we should see `hello` inside the `zod` directory, populated with files.
   - The `sys.kelvin` file specifies the kernel version it's compatible with.

>    - The optional `desk.ship` file specifies the original publisher of this desk. Since we merged this from `%garden` its currently set to `~mister-dister-dozzod-dozzod`. Let's change this to our fake ship `~zod`. In practice you'll set this to the ID you want to be seen as the publisher (likely your own Urbit ID). Currently there is no verification that makes sure publishers are honest about this, but eventually there will be.

Reviewer 4: Its (after %garden) should be it's.

   - ```bash
     echo "~zod" > ~/urbit-dev/zod/hello/desk.ship
     ```

>  - If we had Gall agents in this desk which should be automatically started when the desk is installed, we'd add them to a hoon list in the `desk.bill` file. From our `%garden` merge we currently have this:

Reviewer 4: This sentence is confusing. I suggest clarifying or adding a comma before which.

   - ```bash
     :~  %docket
         %treaty
         %hark-store
         %hark-system-hook
         %settings-store
     ==
     ```
 - Since we're not adding any agents in this example, we'll remove the `desk.bill` file and some other garden specific directories.
   - ```bash
       rm -rf ~/urbit-dev/zod/hello/desk.bill app gen ted
     ```
 - If you `ls` you should see the following in `hello`:
   - `desk.docket-0 desk.ship     lib           mar           sur           sys.kelvin`

>  - The final file we need to edit is `desk.docket-0`. This one's more complicated, so we'll need a text editor.

Reviewer 4: Throughout we need to be consistent about contractions, but for this line specifically I suggest editing "one's" to "one is" for clarity.

         color+0x81.88c9
         version+[0 0 1]
         website+'https://urbit.org/docs/userspace/dist/guide'
         license+'MIT'
         base+'hello'
         glob-ames+[~zod 0v0]
         image+'https://media.urbit.org/docs/userspace/dist/wut.svg'
     ==
     ```
  - This is a `docket` file. This file sets various options for the desk with a tile and usually a brower-based front-end of some kind. 
  - It mainly configures the appearance of an app's tile, the source of its glob (bundle of client resources like HTML, JS, and CSS files), and some additional metadata.
    - We've given the app a `%title` of "Hello", which will be displayed on the app tile and will be the name of the app when others browse to install it. 
    - We've given the app tile a `%color` of #8188C9, and specified the URL of an `%image` to display on the tile.
    - The `%base` clause specifies the base URL path for the app. We've specified "hello" so it'll be `http://localhost:80/apps/hello/...` in the browser.
      - **Note**: `%base` is only put into `/apps/[base]` here *because* the app is served as a glob through docket. A `%site` would be free to bind wherever.

>       - **Note**: `:80` is dependent on urbit being able to bind to that default. If it's in use by something else the port will likely be `:8080`, but you can see which one is in use by looking at the output generated in the terminal when you start your ship. Use the one your ship shows, this guide will assume the defaults are working for examples.

Reviewer 4: Comma after "ship shows" should be semicolon.

    - For the glob, we've used a clause of `%glob-ames`, which means the glob will be served from a ship over Ames, as opposed to being served over HTTP with a `%glob-http` clause or having an Eyre binding with a `%site clause`. 
    - Next we've specified `[~zod 0v0]`. Since `~zod` is the fakeship we'll install it on, the `%docket` agent will wait for a local upload of the glob, so we can just specify `0v0` here as it'll get overwritten later.

>     - The `%version` clause specifies the version as a triple of major version, minor version and patch version. The rest is just some additional informative metadata which will be displayed in App Info.

Reviewer 4: Suggest cutting "just some" for concision.

Which after "informative metadata" should be that.

- With that saved we now have everything we need in the `hello` directory.

##### Committing and installing the changes

 - Let's start our fake ~zod again.
   - ```bash
     cd ~/urbit-dev/
     ./urbit zod
     ```
- Commit and install our changes.
  - ```bash
    |commit %hello
    |install our %hello
    ```

>   - You'll see `docket: awaiting manual glob for %hello desk` this is because our `desk.docket-0` files includes a `%glob-ames` clause which specifies our ship as the source, so it's waiting for us to upload the glob.

Reviewer 4: Period or semicolon before "This is because..."

Not sure if files should be singular or not, but if it's plural than includes should be edited to "include".

Which (after "clause") should be that.

- Open a browser and navigate to `http://localhost:80`. To get the login code, enter `+code` into the dojo prompt for your ship. For fake `~zod`s, the code is always `lidlut-tabwed-pillex-ridrup`. You'll see our tile, but it says installing with a spinner due to the missing glob.
- **Note**: Why are we using globs to hold separate files instead of copying everything into Clay?

>   - We do this as a current VM memory limits work around. Large minimized JS blobs in Clay can't be reduced because each one is one line and unique. This means Clay will grow in size with every JavaScript file permenantly. The decision was made to have the JS remembered by the userspace application instead of writing to clay.

Reviewer 4: Consider editing second sentence to "...can't be reduced because one is a single line and unique" (the repetition of "one" is confusing here).

Also "The decision was made" phrasing is somewhat awkward. Can we specify a subject of the decision and lose the passive sentence construction?

 - ```bash
   cd ~
   mkdir hello-glob
   cd hello-glob
   mkdir img
   curl https://media.urbit.org/docs/userspace/dist/pot.svg --output img/pot.svg
   ```
- We've grabbed an image to use in our "Hello world!" page. The next thing we need to add is an `index.html` file in the root of the folder. The `index.html` file is mandatory; it's what will be loaded when the app's tile is clicked. Let's open our preferred editor and create it.
  - ```bash
    nano index.html
    ```
- Paste the following into the editor and save it.
  - ```html
    <!DOCTYPE html>
    <html>
      <head>
        <title>Hello World</title>
        <style>body { text-align: center; }</style>
      </head>
      <body>
        <h1>Hello World!</h1>
        <img src="img/pot.svg" alt="pot" width="219" height="196" />
      </body>
    </html>
    ```

##### Upload to glob

We can now create a glob from the directory. To do so, navigate to `http://localhost:80/docket/upload` in the browser. This will bring up the `%docket` app's Globulator tool.

> - Simply select the `hello` desk from the drop-down, click `Choose file` and select the `hello-glob` folder in the the file browser, then hit `glob!`.

Reviewer 4: Extra the before "file browser"

 - ```bash
   :treaty|publish %hello
   ```
   - **Note**: For desks without a docket file (and therefore without a tile and glob), treaty can't be used. Instead you can make the desk public with `|public %desk-name`.
   - **Note**: Desks aren't publicly readable by default, `:treaty|publish` publishes the metadata and also makes the desk publicly readable, like `|public`.

##### Remote install

Let's spin up another fake ship so we can try to install our newly published app.

 - ```bash
   cd ~/urbit-dev
   ./urbit -F bus
   ```
   - Note: For desks without a docket file (and therefore without a tile and glob), users cannot install them through the web interface. Instead remote users can install it from the dojo with `|install ~our-ship %desk-name`.
- In the browser, navigate to http://localhost:8081 and login with `~bus`'s code which can be found by running `+code` in the dojo for `bus`. In this case, the default is `riddec-bicrym-ridlev-pocsef`. Next, type `~zod/` in the search bar, and it should pop up a list of `~zod`'s published apps, which in this case is our `Hello` app.
- When we click on the app, it'll show some of the information from the clauses in the docket file.

> - Click `Get App` and it'll ask as if we want to install it.

Reviewer 4: Add period after 'Get App' then cut and to make new sentence.

> - Finally, click `Get "Hello"` and it'll be installed as a tile on `~bus` which can then be opened.

Reviewer 4: Add period after 'Get "Hello"' then cut "and" to make a new sentence.

##### Summary and links

You've done it! You now have a basic understanding of how to create local fake ships to develop with, how to create a basic glob for a front-end, and how to publish your application to the network.

From here you can use your local environment to learn Hoon and then build and publish applications to the network.

There's a lot of detail to get into from here, but the workflow above will give you the basics to explore further.

 - More about [Docket Files](https://urbit.org/docs/userspace/dist/docket)
 - More about [Glob](https://urbit.org/docs/userspace/dist/glob)
 - Learn [Hoon](https://urbit.org/docs/hoon/overview)

### Advanced tips and tricks

#### Faster fake ship booting

> A common part of the development workflow is deleting and creating fake ships. The couple of minutes this takes slows down development. Instead we can create a backup copy of a freshly created fake ship. This way when you want to delete some fake ship you've been working on and reset it, it can be done instantly rather than having to wait for a new one to be created.

Reviewer 4: Second it after "reset it" could be specified to something like "the process" to eliminate repetition and confusion. Add semicolon after first it.

# Tlon Dev Guide

## What is this guide?

> This is Tlon's guide for new developers that want to build applications on Urbit. The goal of this guide is to take you from a fresh machine to a working development environment for Urbit applications. This guide assumes macOS, but Linux is also supported (see the Urbit foundation docs for Linux support) It's important this guide remains clear and correct, if you find something broken please let us know at support@tlon.io

Reviewer 5: Memo to self and @tinnus-napbus to make sure we have this in good nick.

## Pre-requisites

### Homebrew

- Homebrew is a package manager for macOS, we'll need it installed in order to install `git-lfs`.

> - ```/bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)"```

Reviewer: There might be an argument for just linking to brew.sh instead, so that we don't risk giving outdated installation instructions.

Author: I think that's a reasonable argument generally, but in this specific case that install line has remain unchanged for many years (iirc) and it's worth it I think not to make the user leave the page.

### Git LFS

- A package that enables git to work with large files, we'll install this via brew.
- ```brew install git-lfs```

### Visual Studio Code

>  - While not technically a prerequisite, there's a hoon syntax highlighting plugin available and this is the editor we recommend: https://code.visualstudio.com/

Reviewer: ```suggestion
 - While not technically a prerequisite, there's a hoon syntax highlighting plugin available and this is the editor we recommend [Visual Studio Code](https://code.visualstudio.com/).
```

 - For the small text file edits in this guide, we'll simply use `nano`.

>  ### Urbit Git

Reviewer: ```suggestion
### The Urbit Git repository
```

  - We're going to need a local copy of the `urbit` code.
     - ```bash
        git clone https://github.com/urbit/urbit ~/urbit-git
       ```

>   - Pull down the git lfs files.  

Reviewer: ```suggestion
  - We're going to need a local copy of the Urbit source code.
     - ```bash
       git clone https://github.com/urbit/urbit ~/urbit-git
       ```
  - Pull down the files stored using git lfs.
```

Or something similar. Also, since I found trailing spaces here, consider configuring your editor to delete those on-save.

    - ```bash
      cd ~/urbit-git
      git lfs install
      git lfs pull
      ```

> ### Urbit Install

Reviewer: ```suggestion
### The Urbit binary
```

Also kind of surprising how we get the source code, but then just download a pre-built binary, especially since we didn't elaborate on why we need the source code in the first place.

I guess having the reader install nix just to build the binary locally is a bit much, but this section would be much nicer if we could replace it with just `cd pkg/urbit && build` or w/e.

Author: Yeah fair flag, I figured more can go wrong on local build - but maybe not? Going to leave this one for v0, but can change if we want to later.

 - First we need to create a directory for the urbit binary, we'll create this in your home directory.
   - ```bash
     mkdir ~/urbit-dev
     ```
 - Next we're going to pull down the latest urbit binary from urbit.org from the newly created directory.
    - ```bash
      cd ~/urbit-dev
      curl -JLO https://urbit.org/install/mac/latest
      ```
  - Now we'll untar the contents of the download into our newly created directory.
    - ```bash 
      tar -zxvf ./darwin.tgz --strip=1
      ```
  - With the urbit binary installed we're ready for the basic dev workflow.

## Development Basics

### Fake Ships

#### What is a "fake" ship?

>  - For local development we'll create a 'fake' ship, this is a local ship that is not connected to the urbit network. This means we don't need a real ID (eg comet, planet) and can delete and recreate the ship whenever we want to reset things.

Reviewer: ```suggestion
 - For local development we'll create a "fake" ship, this is a local ship that is not connected to the Urbit network. This means we don't need a real ID (eg comet, planet) and can delete and recreate the ship whenever we want to reset things.
```

For consistency with the quite style used in the headings.

#### What is a "pill"?

>  - A pill file contains the boot sequence, if it's not specified the livenet pill is used by default. 

Reviewer: ```suggestion
 - A pill file contains a sequence of events that initializes an urbit. If no pill is specified, the livenet pill is downloaded instead. 
```

Might want to rephrase to keep "boot" in there somehow, but explaining this correctly while staying concise is tough...

Reviewer 5: Are we using "an urbit" instead of "a ship" in terminology?

>  - Since OS updates propogate through the network, it's possible the default pill file could be slightly out of date.

Reviewer: ```suggestion
 - Since OS updates propagate through the network, it's possible the default pill file could be slightly out of date.
```

>  - We can manually get the most up to date pill file for local development by downloading it and using it when starting our fake ship.

Reviewer: This is... not right? The "downloading part", that is. There is no place to get the "latest" pill from other than either the repo or bootstrap.urbit.org (the default), and at the time of writing those are usually equivalent, though the one in the repo might be newer.

Maybe that's what you meant, that it "downloaded" with the rest of the repo? Worth rephrasing.

Reviewer 5: Could talk about how to get a comet and make a pill, but not really a basic quick-n-dirty intro.

Author: ack - yeah I meant downloaded with the repo, will clarify.

   - For fake development ships not talking to the regular network this works well.
   - We don't do this for 'real' ships on the actual network because if you're using a pill that's newer than what your sponsor has (the ship that you get updates from) your sponsor will think you're on an older version.
   - When this happens your sponsor will attempt to upgrade you, but will actually be downgrading you to their verison.

>    - There are compiled and uncompiled pills. They are differentiated by a naming convention that adds `-brass` to the compiled pill name. For example, `multi.pill` is the name of a compiled pill and the name of its uncompiled version is `multi-brass.pill`. For the purposes of this guide we'll use compiled pills because they boot faster. An uncompiled pill lets you audit the source code if you need to.

Reviewer: This whole section on pills feels a little too in the weeds for a "getting started" dev guide. Pill freshness _might_ be relevant for figuring out whether you're developing against the right version of the stdlib or some app or w/e, but I don't see the various pill types being relevant at all.

Probably condense down to "we use `solid.pill` because it contains the pre-compiled system and so helps us boot fast", or just remove entirely. I guess the `solid` vs `multi` distinction may be worth talking about.

Author: Yeah, I think this is worth refactoring. It's a tradeoff between trying to explain what this pill is that's mentioned and not getting too detailed that it's confusing. I'll try to reduce it.

Author: I'm not sure I know the solid vs. multi distinction? I grabbed `multi.pill` because it was in the other guide I had been referencing parts of.

Author: guessing multi-pill contains stuff for all the desks and solid is just for base?

 - In summary, if you're using a fake ship -> get the latest pill, if you're starting a real ship don't worry about pills.

#### Creating a fake ship

 - We can create a local fake ship with the below command. This will create a fake ship named zod in our `urbit-dev` directory using the newest `multi.pill` from the urbit code we cloned earlier. This will take a couple of minutes and generate some output in the terminal.
   - ```bash
     ./urbit -F zod -B ~/urbit-git/bin/multi.pill

>      ```

Reviewer: May or may not be interesting to also introduce `-c zod-dev` and saying that creates the fake ship in a folder named `zod-dev`. Shows that pier folder names aren't strictly tied to the ship in it, can be used for organization. Could also just be not relevant enough though.

 - When it's finished you'll see a command prompt like `~zod:dojo>` - this is the urbit dojo where you can enter and run urbit commands.

>  - To exit the dojo type `|exit` or `ctrl-D`

Reviewer: ```suggestion
 - To exit the dojo type `|exit` or `ctrl-D`.
```

#### Fake ship networking

>  - Fake ships on the same machine can automatically talk to each other. Let's create a second fake ship named `~bus` and confirm this.

Reviewer 5: There are rules about routing tho, correct?  I.e. you need to make sure two planets are from same star.

Reviewer 6: hmm yeah apparently, I didn't realise that

Author: added a note

   - ```bash
     ./urbit -F bus -B ~/urbit-git/bin/multi.pill
     ```
 - Now you'll have two local ships, `~zod` and `~bus`.
 - From the newly created `~bus` dojo enter `|hi ~zod` and you'll see that urbit's networking finds the first fake ship we created.

>  - Exit the `~bus` dojo with `|exit` or `ctrl-D`

Reviewer: Note that we just instructed the user to shut down the zod, so the `|hi` may not arrive. (;

```suggestion
 - From the newly created `~bus` dojo enter `|hi ~zod` and you'll see that urbit's networking finds the first fake ship we created.
 - Exit the `~bus` dojo with `|exit` or `ctrl-D`.
```

(And other places. We should be consistent about this.)

### Desks

#### What is a desk?

> A desk is an independently revision-controlled branch of a ship that uses the Clay filesystem. Each desk contains its own apps, mark definitions, files, etc. You can think of a desk like a git branch you're working in.

Reviewer: Do we assume the reader knows what a mark is? If not, might be better to substitute with the more generic "type".

```suggestion
A desk is an independently revision-controlled branch of a ship that uses the Clay filesystem. Each desk contains its own apps, type definitions, files, etc. You can think of a desk like a git branch you're working in.
```

> A basic ship has a `%base`, `%garden`, `%landscape`, `%bitcoin`, and `webterm` desk (we mounted these earlier). The `%base` desk has the kernel and base system software, while other desks are each for different apps. 

Reviewer: > we mounted these earlier

We will mount these later!

Also, this assumes you're booting with the multipill, which I wouldn't really describe as "a basic ship". The most-basic ship only needs a `%base` desk.

Author: I removed that parenthetical since the mounting happens after now anyway. I guess standard may be better than basic? I left it in for now.

A desk is a series of numbered commits, the most recent of which represents the current state of the desk. A commit is composed of an absolute time when it was created, a list of zero or more parents, and a map from paths to data.

> For building applications to distribute on urbit, we'll want to work on a new desk.

Reviewer: ```suggestion
For building applications to distribute on Urbit, we'll want to work on a new desk.
```

#### Desk Basics

> ##### Mounting and Unmounting Desks

Reviewer: ```suggestion
##### Mounting and unmounting Desks
```

Title-case is for click-bait. Some other headings should be updated as well.

 - Mounting a desk allows you to add files to it (it creates a directory you can see on the host filesystem). Run these commands in the dojo.
   - ```bash
     |mount %base
     ```
 - You can similarly unmount a desk
   - ```bash
     |unmount %base
     ```

##### Setting Up a New Desk

>  - To create a new desk we'll need to merge from an existing one, think of this like creating a new branch.

Reviewer: ftr, we'll need to update this once urbit/urbit#5360 lands.

>  - Let's create a new `hello` desk by merging from `%garden` (`garden` is just an old name for the application now referred to as `landscape`).

Reviewer: Consider using a desk with less baggage. (^:

Author: haha well this is one they'll have to learn sometime

 - **Note**: Thinking about dependencies:
   - When we merge from `%garden` we start with its base files in our application.
   - Since we're copying these files over, its possible an API change could happen in `urbit-git` while we're building our app.

>    - It's good practice to run the merge command at the start of development and prior to deployment (after a fresh pull of `urbit-git`) to make sure it works with the up to date system changes.

Reviewer: This section feels a bit confusing somehow. Would probably help to place it after the merge instructions, at least.

  - Let's start up our fake zod.
    - ```bash
      cd ~/urbit-dev
      ./urbit zod
      ```
  - Create our desk in the dojo by merging from `%garden`
    - ```bash
      |merge %hello our %garden
      ```
  - Mount `%hello` so we can access it from the host filesystem.
    - ```bash
      |mount %hello
      ```

>   - Exit the Dojo with `|exit` or `ctrl-D` 

Reviewer: Why though? I guess we're doing all this in a single terminal window? Might be worth to have the user just open two windows instead.

Reviewer 5: +1 to two windows.

Also consider prefacing Bash commands with `$` and Dojo commands with `dojo>`.

Author: I avoid the preface just because it can make copy pasting commands a pain.

Can do two windows, but it increases the risk of them accidentally running the same ship at the same time though (which I guess matters less for this? should also get caught by the lock I think).

 - Now we should see `hello` inside the `zod` directory, populated with files.
   - The `sys.kelvin` file specifies the kernel version it's compatible with.

>    - The optional `desk.ship` file specifies the original publisher of this desk. Since we merged this from `%garden` its currently set to `~mister-dister-dozzod-dozzod`. Let's change this to our fake ship `~zod`, in practice you'll set this to the ID you want to be seen as the publisher (likely your urbit ID). Currently there is no verification that makes sure publishers are honest about this, but eventually there will be.

Reviewer: ```suggestion
   - The optional `desk.ship` file specifies the original publisher of this desk. Since we merged this from `%garden` its currently set to `~mister-dister-dozzod-dozzod`. Let's change this to our fake ship `~zod`. In practice you'll set this to the ID you want to be seen as the publisher (likely your own Urbit ID). Currently there is no verification that makes sure publishers are honest about this, but eventually there will be.
```

   - ```bash
     echo "~zod" > ~/urbit-dev/zod/hello/desk.ship
     ```
 - If we had Gall agents in this desk which should be automatically started when the desk is installed, we'd add them to a hoon list in the `desk.bill` file. From our `%garden` merge we currently have this:
   - ```bash
      :~  %docket
          %treaty
          %hark-store
          %hark-system-hook
          %settings-store
      ==
     ```
 - Since we're not adding any agents in this example, we'll remove the `desk.bill` file and some other garden specific directories.
   - ```bash
       rm ~/urbit-dev/zod/hello/desk.bill app gen ted

>      ```

Reviewer: ```suggestion
   - ```bash
     :~  %docket
         %treaty
         %hark-store
         %hark-system-hook
         %settings-store
     ==
     ```
 - Since we're not adding any agents in this example, we'll remove the `desk.bill` file and some other garden specific directories.
   - ```bash
     rm ~/urbit-dev/zod/hello/desk.bill app gen ted
     ```
```

Isn't this `rm` going to want a `-r` for those directories though?

 - If you `ls` you should see the following in `hello`:
   - `desk.docket-0 desk.ship     lib           mar           sur           sys.kelvin`
 - The final file we need to edit is `desk.docket-0`. This one's more complicated, so we'll need a text editor.
   - ```bash
     nano desk.docket-0
     ```
   - In the text editor you'll see the following that came in from `%garden`:
   - ```bash
        :~ title+'System'
           info+'An app launcher for Urbit.'
           color+0xee.5432
           version+[1 0 2]
           website+'https://tlon.io'
           license+'MIT'
           base+'grid'
           glob-http+['https://bootstrap.urbit.org/glob-0v4.64ana.19ug9.ik7l6.og080.68ce4.glob' 0v4.64ana.19ug9.ik7l6.og080.68ce4]

>         ==

Reviewer: ```suggestion
   - ```bash
     :~  title+'System'
         info+'An app launcher for Urbit.'
         color+0xee.5432
         version+[1 0 2]
         website+'https://tlon.io'
         license+'MIT'
         base+'grid'
         glob-http+['https://bootstrap.urbit.org/glob-0v4.64ana.19ug9.ik7l6.og080.68ce4.glob' 0v4.64ana.19ug9.ik7l6.og080.68ce4]
     ==
```

     ```
   - We're going to modify this for our app to the below:
   - ```bash
     :~  title+'Hello'
         info+'A simple hello world app.'
         color+0x81.88c9
         version+[0 0 1]
         website+'https://urbit.org/docs/userspace/dist/guide'
         license+'MIT'
         base+'hello'
         glob-ames+[~zod 0v0]
         image+'https://media.urbit.org/docs/userspace/dist/wut.svg'
     ==
     ```
  - This is a `docket` file. This file sets various options for the desk with a tile and usually a brower-based front-end of some kind. 
  - It mainly configures the appearance of an app's tile, the source of its glob (bundle of client resources like HTML, JS, and CSS files), and some additional metadata.
    - We've given the app a `%title` of "Hello", which will be displayed on the app tile and will be the name of the app when others browse to install it. 

>     - We've given the app tile a `%color` of #8188C9, and specified the URL of an `%image` to display on the tile.

Reviewer: Honestly might be better to just say "given the app a title of "Hello"" instead, same for the others. The reader may not know about `a+b` sugar and just assume all names are `%names` or w/e. It's overkill.

Also we mention "clause" in a couple places but never explain what that means.

    - The `%base` clause specifies the base URL path for the app. We've specified "hello" so it'll be `http://localhost:80/apps/hello/...` in the browser.

>       - **Note**: `%base` is only put into `/apps/[base]` here *because* the app is served as a glob through docket. A `%site` would be free to bind whereever.

Reviewer: ```suggestion
      - **Note**: `%base` is only put into `/apps/[base]` here *because* the app is served as a glob through docket. A `%site` would be free to bind wherever.
```

      - **Note**: `:80` is dependent on urbit being able to bind to that default. If it's in use by something else the port will likely be `:8080`, but you can see which one is in use by looking at the output generated in the terminal when you start your ship. Use the one your ship shows, this guide will assume the defaults are working for examples.
    - For the glob, we've used a clause of `%glob-ames`, which means the glob will be served from a ship over Ames, as opposed to being served over HTTP with a `%glob-http` clause or having an Eyre binding with a `%site clause`. 

>     - Next we've specified `[~zod 0v0]`. Since `~zod` is the fakeship we'll install it on, the `%docket` agent will wait for a separate upload of the glob, so we can just specify `0v0` here as it'll get overwritten later.

Reviewer: ```suggestion
    - Next we've specified `[~zod 0v0]`. Since `~zod` is the fakeship we'll install it on, the `%docket` agent will wait for a local upload of the glob, so we can just specify `0v0` here as it'll get overwritten later.
```

    - The `%version` clause specifies the version as a triple of major version, minor version and patch version. The rest is just some additional informative metadata which will be displayed in App Info.
- With that saved we now have everything we need in the `hello` directory.

##### Committing and Installing The Changes

>  - Let's start up fake zod.

Reviewer: ```suggestion
 - Let's start our fake ~zod again.
```

   - ```bash
     cd ~/urbit-dev/
     ./urbit zod
     ```
- Commit and install our changes.
  - ```bash
    |commit %hello
    |install our %hello
    ```
  - You'll see `docket: awaiting manual glob for %hello desk` this is because our `desk.docket-0` files includes a `%glob-ames` clause which specifies our ship as the source, so it's waiting for us to upload the glob.

> - Open a browser and navigate to `http://localhost:80`. To get the login code, enter `+code` into the dojo prompt for your ship. In this case, the default is `lidlut-tabwed-pillex-ridrup`. You'll see our tile, but it says installing with a spinner due to the missing glob.

Reviewer: ```suggestion
- Open a browser and navigate to `http://localhost:80`. To get the login code, enter `+code` into the dojo prompt for your ship. For fake ~zods, the code is always `lidlut-tabwed-pillex-ridrup`. You'll see our tile, but it says installing with a spinner due to the missing glob.
```

##### Create Files for Glob

> We'll now create the files for the glob. We'll use a very simple static HTML page that just displayes "Hello World!" and an image. Typically we'd have a more complex JS web app that talked to apps on our ship through Eyre's channel system, but for the sake of simplicity we'll forgo that. Let's hop back in the Unix terminal.

Reviewer: ```suggestion
We'll now create the files for the glob. We'll use a very simple static HTML page that just displays "Hello World!" and an image. Typically we'd have a more complex JS web app that talked to apps on our ship through Eyre's channel system, but for the sake of simplicity we'll forgo that. Let's hop back in the Unix terminal.
```

Also what is "Eyre's channel system"? We don't mention it further. Should we link to docs about it? Alternatively just not bring it up.

Reviewer 5: I'd avoid mentioning Eyre here.  Perhaps:

> Typically we'd have a more complex JS web app that talked to apps on our ship through the server API

 - ```bash
   cd ~
   mkdir hello-glob
   cd hello-glob
   mkdir img
   curl https://media.urbit.org/docs/userspace/dist/pot.svg --output img/pot.svg
   ```
- We've grabbed an image to use in our "Hello world!" page. The next thing we need to add is an `index.html` file in the root of the folder. The `index.html` file is mandatory; it's what will be loaded when the app's tile is clicked. Let's open our preferred editor and create it.
  - ```bash
    nano index.html
    ```
- Paste the following into the editor and save it.
  - ```html
    <!DOCTYPE html>
    <html>
      <head>
        <title>Hello World</title>
        <style>body { text-align: center; }</style>
      </head>
      <body>
        <h1>Hello World!</h1>
        <img src="img/pot.svg" alt="pot" width="219" height="196" />
      </body>
    </html>
    ```
- **Note**: Why are we using globs to hold separate files instead of copying everything into Clay?

>   - We do this as a current loom limits work around.

Reviewer 5: Recommend not mentioning `loom`, just `VM memory limits`.

Reviewer 5: This and the next three lines should be a single item or paragraph rather than separate entries.

Author: Thanks - I should have time to refactor all of this today and push an update

>   - Large minimized js blobs in Clay can't be reduced because each one is one line and unique.
>   - This means Clay will grow in size with every javascript file permenantly.
>   - The decision was made to have the js remembered by the userspace application instead of writing to clay.

Reviewer: ```suggestion
  - Large minimized JS blobs in Clay can't be reduced because each one is one line and unique.
  - This means Clay will grow in size with every JavaScript file permenantly.
  - The decision was made to have the JS remembered by the userspace application instead of writing to clay.
```
This section feels like it comes kind of in the middle of things. I would've expected it earlier (when explaining the `glob-ames` line), or at the end of this once we've finished doing the thing this describes.

  - This prevents Clay from growing permenantly in size.

>   - In the future this will be resolved in two ways: hoon front-end development and Clay tombstones (so we can reduce the size of older js blobs we no longer need).

Reviewer: ```suggestion
  - In the future this will be resolved in two ways: Hoon front-end development and Clay tombstones (so we can fully forget older JS blobs we no longer need).
```

  - For the purposes of this guide you don't need to worry about this, it's just helpful context for advanced users.

##### Upload to Glob

> We can now create a glob from the directory. To do so, navigate to http://localhost:80/docket/upload in the browser. This will bring up the `%docket` app's Globulator tool.

Reviewer: ```suggestion
We can now create a glob from the directory. To do so, navigate to `http://localhost:80/docket/upload` in the browser. This will bring up the `%docket` app's Globulator tool.
```

Here and below.

- Simply select the `hello` desk from the drop-down, click `Choose file` and select the `hello-glob` folder in the the file browser, then hit `glob!`.
- Now if we return to our ship's homescreen, we should see the tile looks as we specified in the docket file: http://localhost:80

> - And if we click on the tile, it'll load the index.html in our glob.

Reviewer: ```suggestion
- And if we click on the tile, it'll load the `index.html` in our glob.
```

- Our app is working!

##### Publish

The final step is publishing our desk with the `%treaty` agent so others can install it. To do this, there's a simple command in the dojo.
 - ```bash
   :treaty|publish %hello
   ```
   - **Note**: For desks without a docket file (and therefore without a tile and glob), treaty can't be used. Instead you can make the desk public with `|public %desk-name`.
   - **Note**: Desks aren't publicly readable by default, `:treaty|publish` publishes the metadata and also makes the desk publicly readable, like `|public`.

##### Remote Install

Let's spin up another fake ship so we can try to install our newly published app.

 - ```bash
   cd ~/urbit-dev
   ./urbit -F bus
   ```
   - Note: For desks without a docket file (and therefore without a tile and glob), users cannot install them through the web interface. Instead remote users can install it from the dojo with `|install ~our-ship %desk-name`.

> - In the browser, navigate to http://localhost:8081 and login with `~bus`'s code which can be found by running `+code` in the dojo for `bus`. In this case, the default is `riddec-bicrym-ridlev-pocsef`. Next, type `~zod/` in the search bar, and it should pop up a list of `~zod`'s published apps, which in this case is our `Hello` app.

Reviewer: ```suggestion
- In the browser, navigate to http://localhost:8081 and login with ~bus's code which can be found by running `+code` in its dojo. For fake ~busses, the code is always `riddec-bicrym-ridlev-pocsef`. Next, type `~zod/` in the search bar, and it should pop up a list of ~zod's published apps, which in this case is our `Hello` app.
```

None of the rest of the document `monospaces` the ship names, I don't think.

Same change below, and in the tips & tricks section.

Reviewer 5: Neither here nor there for this doc, but my preference is to monospace ship names if the style doesn't change the background color, so they still read smoothly.

Author: Looked like half and half, defaulted to monospacing them for now.

- When we click on the app, it'll show some of the information from the clauses in the docket file.
- Click `Get App` and it'll ask as if we want to install it.
- Finally, click `Get "Hello"` and it'll be installed as a tile on `~bus` which can then be opened.

##### Summary and Links

You've done it! You now have a basic understanding of how to create local fake ships to develop with, how to create a basic glob for a front-end, and how to publish your application to the network.

> From here you can use your local environment to learn hoon and then build and publish applications to the network.

Reviewer: ```suggestion
From here you can use your local environment to learn Hoon and then build and publish applications to the network.
```

 - Learn [Hoon](https://urbit.org/docs/hoon/overview)

### Advanced Tips and Tricks

#### Faster fake ship booting

A common part of the development workflow is deleting and creating fake ships. The couple of minutes this takes slows down development. Instead we can create a backup copy of a freshly created fake ship. This way when you want to delete some fake ship you've been working on and reset it, it can be done instantly rather than having to wait for a new one to be created.
 - We'll do this now by backing up the fake `~zod` we just created.
 - First start up `~zod` from our `~/urbit-dev` directory. Notice that we don't need the `-F` argument this time because `~zod` is already created.
   - ```bash
     ./urbit zod
     ```
 - Next mount every desk (this is so we don't have to do this every time after resetting our fake ship). Each of these is one dojo command.
   - ```bash
     |mount %base
     |mount %garden
     |mount %landscape
     |mount %bitcoin
     |mount %webterm
     ```
 - Next shutdown `~zod` with `|exit` or `ctrl-D`
 - Now we're going to make a copy of our fake ship `~zod` and call it `zod.bk`
   - ```bash
     cp -r zod zod.bk
     ```

>  - Then you'll dev with `~zod` like normal, but when you're ready to reset it to a fresh state, instead of creating a new one we'll just delete `~zod` and copy the backup over.

Reviewer 5: Could also get into just rm/cp `.urb` but perhaps too complicated here.

Reviewer 5: Depends on what they're likely to want to reset, I suppose.
